File: IguideME.Web/Models/Tile.cs
using System.Collections.Generic;
using IguideME.Web.Models.App;
using Newtonsoft.Json;

namespace IguideME.Web.Models
{

	public class TilesGrades
	{

		[JsonProperty(PropertyName = "tile_id")]
		public int tile_id { get; set; }
		[JsonProperty(PropertyName = "grade")]
		public double grade { get; set; }
		[JsonProperty(PropertyName = "max")]
		public double max { get; set; }
		public TilesGrades(int id, double grade, double max)
		{
			this.tile_id = id;
			this.grade = grade;
			this.max = max;

		}
	}

	public class UserTileGrades
	{
		[JsonProperty(PropertyName = "userID")]
		public string userID { get; set; }

		[JsonProperty(PropertyName = "tile_grades")]
		public TilesGrades[] tile_grades { get; set; }

		public UserTileGrades(string userID, TilesGrades[] grades)
		{
			this.userID = userID;
			this.tile_grades = grades;
		}

	}

	public enum TileType
	{
		assignments,
		discussions,
		learning_outcomes
	}

	public class Tile
	{
		[JsonProperty(PropertyName = "id")]
		public int ID { get; private set; }

		[JsonProperty(PropertyName = "group_id")]
		public int GroupID { get; set; }

		[JsonProperty(PropertyName = "title")]
		public string Title { get; set; }

		[JsonProperty(PropertyName = "position")]
		public int Order { get; set; }

		[JsonProperty(PropertyName = "type")]
		public TileType Type { get; set; }

		[JsonProperty(PropertyName = "weight")]
		public double Weight { get; set; }

		[JsonProperty(PropertyName = "gradingType")]
		public AppGradingType GradingType { get; set; }

		[JsonProperty(PropertyName = "alt")]
		public bool Alt { get; set; }

		[JsonProperty(PropertyName = "visible")]
		public bool Visible { get; set; }

		[JsonProperty(PropertyName = "notifications")]
		public bool Notifications { get; set; }

		[JsonProperty(PropertyName = "entries")]
		public List<TileEntry> Entries { get; set; }

		public Tile(
			int id,
			int groupID,
			string title,
			int order,
			TileType type,
			double weight,
			AppGradingType gradingType,
			bool alt,
			bool visible,
			bool notifications
		)
		{
			this.ID = id;
			this.GroupID = groupID;
			this.Title = title;
			this.Order = order;
			this.Type = type;
			this.Weight = weight;
			this.GradingType = gradingType;
			this.Alt = alt;
			this.Visible = visible;
			this.Notifications = notifications;
			this.Entries = new();
		}
	}

	public class TileEntry
	{
		[JsonProperty(PropertyName = "tile_id")]
		public int TileID { get; set; }

		[JsonProperty(PropertyName = "title")]
		public string Title { get; set; }

		[JsonProperty(PropertyName = "content_id")]
		public int ContentID { get; set; }

		[JsonProperty(PropertyName = "weight")]
		public double Weight { get; set; }

		public TileEntry(int tileID, int ContentID, string title, double weight)
		{
			this.TileID = tileID;
			this.ContentID = ContentID;
			this.Title = title;
			this.Weight = weight;
		}
	}

	public class TileGrades
	{
		[JsonProperty(PropertyName = "tile_id")]
		public int TileID { get; set; }

		[JsonProperty(PropertyName = "grades")]
		public AppGrades Grades { get; set; }


		public TileGrades(int tileID, AppGrades grades)
		{
			this.TileID = tileID;
			this.Grades = grades;
		}
	}
}

File: IguideME.Web/Frontend/src/types/tile.ts
import { type Grades, type GradingType } from './grades';

export type ViewType = 'graph' | 'grades';

export interface LayoutColumn {
  id: number;
  width: number;
  position: number;
  groups: number[];
}

export interface TileGroup {
  id: number;
  title: string;
  position: number;
}

export interface Tile {
  id: number;
  group_id: number;
  title: string;
  position: number;
  type: TileType;
  weight: number;
  alt: boolean;
  visible: boolean;
  notifications: boolean;
  gradingType: GradingType;
  entries: TileEntry[];
}

export enum TileType {
  assignments,
  discussions,
  learning_outcomes,
  external,
}

export interface TileEntry {
  tile_id: number;
  title: string;
  html_url: string;
  published: PublishedStatus;
  weight: number;
  content_id: number;
}

export enum PublishedStatus {
  NotPublished,
  LMSPublished,
  ExternalData,
}

export interface Assignment {
  id: number;
  course_id: number;
  title: string;
  html_url: string;
  published: PublishedStatus;
  muted: boolean;
  due_date: number;
  max_grade: number;
  grading_type: GradingType;
}

export interface DiscussionTopic {
  id: number;
  parent_id: number;
  course_id: number;
  title: string;
  author: string;
  html_url: string;
  date: number;
  message: string;
  grades?: Grades;
}

export interface DiscussionEntry {
  id: number;
  discussion_id: number;
  parent_id: number;
  course_id: number;
  author: string;
  date: number;
  message: string;
}

export interface LearningGoal {
  id: number;
  title: string;
  requirements: GoalRequirement[];
  results?: boolean[];
}

export enum LogicalExpression {
  Less,
  LessEqual,
  Equal,
  GreaterEqual,
  Greater,
  NotEqual,
}

export const printLogicalExpression = (expression: LogicalExpression): string => {
  switch (expression) {
    case LogicalExpression.NotEqual:
      return '≠';
    case LogicalExpression.Less:
      return '<';
    case LogicalExpression.LessEqual:
      return '≤';
    case LogicalExpression.Equal:
      return '=';
    case LogicalExpression.GreaterEqual:
      return '≥';
    case LogicalExpression.Greater:
      return '>';
  }
};

export interface GoalRequirement {
  id: number;
  goal_id: number;
  assignment_id: number;
  value: number;
  expression: LogicalExpression;
}
